course7/homeworks/fred/complex-rpn/main.cc

```diff
@@ -0,0 +1,65 @@
+#include "complex.h"
+#include "calculator.h"
+#include "utils.h"
+
+#include <iostream>
+#include <stdexcept>
+
+void insert(RpnCalculator& calculator, Complex complex) {
+  calculator.insert(complex);
+}
+
+void insert(RpnCalculator& calculator, Operator op) {
+  auto result = calculator.insert(op);
+  std::cout << "Result so far: " << result << std::endl;
+}
+
+int main(int argc, char** args) {
+  std::cout << "Welcome!" << std::endl;
+  std::cout << "Type 'q' to quit" << std::endl << std::endl;
+
+  RpnCalculator calculator;
+
+  std::string input = "";
+  while (true) {
+    std::getline(std::cin, input);
+    if (input == "q")
+      break;
+
+    bool parseFailed = false;
+
+    try {
+      // try to parse a number first
+      auto parseComplexResult = tryParseComplexFromString(input);
+      if (parseComplexResult.first) {
+        insert(calculator, parseComplexResult.second);
+      }
+      else {
+        // if failed, try to parse it as an operator
+        auto parseOperatorResult = tryParseOperatorFromString(input);
+        if (parseOperatorResult.first) {
+          insert(calculator, parseOperatorResult.second);
+        }
+        else {
+          parseFailed = true;
+        }
+      }
+    }
+    catch (const std::runtime_error& e) {
+      std::cout << e.what() << std::endl;
+    }
+
+    if (parseFailed) {
```

Comment:
Here you can see the cost of the 'tryParse' approach: you have, competing methods of reporting errors: by exception and by return value, and you need to have two separate error handlers.